Refactor beatmap set covers into using `ModelBackedDrawable<T>`

frenzibyte opened the pull request:
Closes #5246
Supersedes #5248

Opened a new fresh PR as that previous one is way too old to update (old by half the issue/PR numbers :p, and `lazygit` exploded on me while attempting to do so anyways). 

The minor difference between this and the old one is that in here, I've left the placeholder box where it used to be, as:
 1. it doesn't really make sense for me making the placeholder box a drawable MBD would transition into, instead of always being in the background and make MBD not create anything on `null` beatmap set.
 2. it would get rid of depending on the framework-side PR which seems really weird to have currently.

Note:
 - Now when the cover reloads, it'll appear immediately, ~~that's a known issue for `MBD` which will be fixed separately in https://github.com/ppy/osu-framework/pull/4167~~.
 - [x] Depends on https://github.com/ppy/osu-framework/pull/4167.
 - [x] Depends on framework bump

---

This also fixes a potential exception I encountered (that's why I came back to this), which happens when `UpdateableBeatmapSetCover.BeatmapSet` changes while the previous cover still hasn't passed the load delay, causing the previous cover to attempt constructing a new `BeatmapSetCover` with null beatmap set, and therefore cause the exception.
```
[runtime] 2021-01-19 19:17:15 [error]: System.ArgumentNullException: Value cannot be null. (Parameter 'set')
[runtime] 2021-01-19 19:17:15 [error]: at osu.Game.Beatmaps.Drawables.BeatmapSetCover..ctor(BeatmapSetInfo set, BeatmapSetCoverType type) in /Users/salman/Desktop/osu/osu.Game/Beatmaps/Drawables/BeatmapSetCover.cs:line 20
[runtime] 2021-01-19 19:17:15 [error]: at osu.Game.Beatmaps.Drawables.UpdateableBeatmapSetCover.<updateCover>b__11_0() in /Users/salman/Desktop/osu/osu.Game/Beatmaps/Drawables/UpdateableBeatmapSetCover.cs:line 72
[runtime] 2021-01-19 19:17:15 [error]: at osu.Framework.Graphics.Containers.DelayedLoadWrapper.BeginDelayedLoad()
```

While the obvious fix is to just cancel creating a cover if beatmap set became null, or remove the previous wrapper immediately when beatmap set changes if it still didn't finish loading, I have realized `ModelBackedDrawable` handles this issue pretty well ([by doing the latter internally](https://github.com/ppy/osu-framework/blob/ba7e1852a4c79f74404baf5ca3e5712a90ff2a66/osu.Framework/Graphics/Containers/ModelBackedDrawable.cs#L107-L112)), therefore I decided to just let `UpdateableBeatmapSetCover` use MBD already, with as straightforward changes as possible.

And for that, this resolves https://github.com/ppy/osu/issues/12631 as well.

frenzibyte committed:
Refactor beatmap set covers into using `ModelBackedDrawable<T>`

frenzibyte committed:
Add test scene with failing test case

frenzibyte committed:
Use old-style Enum.GetValues method for older than .NET 5

frenzibyte committed:
Fix forgotten modulo

bdach commented on osu.Game.Tests/Visual/UserInterface/TestSceneUpdateableBeatmapSetCover.cs:
```diff
@@ -57,11 +57,15 @@ public void TestUnloadAndReload()
                 var coverTypes = Enum.GetValues<BeatmapSetCoverType>();
+#else
+                var coverTypes = Enum.GetValues(typeof(BeatmapSetCoverType)).Cast<BeatmapSetCoverType>();
+#endif
```
i'd probably advise *not* using conditional compilation symbols unless absolutely necessary.

frenzibyte replied:
So in these cases, always going old-style is preferrable?

bdach replied:
if feasible, yes. conditional compilation symbols are annoying to work with, as they work at source-level (as in, if a conditional is not active, the code within is fully dead, and won't get syntax highlighting/compile checking; think of it as if it just wasn't there) - so now you have to build for all targets to make sure the project always builds.

frenzibyte committed:
Remove conditional compilation symbols for such case

peppy committed:
Merge branch 'master' into mbd-beatmap-set-cover

bdach committed:
Fix possible multiple enumeration

bdach committed:
Merge branch 'master' into mbd-beatmap-set-cover

peppy commented:
Something looks wrong:


https://user-images.githubusercontent.com/191335/116963580-34020280-ace4-11eb-9f5b-7cd1a348bfed.mp4

peppy reviewed (changes requested):
As posted.

bdach committed:
Add visual test case for crossfade behaviour

bdach committed:
Fix both covers showing if cover is not fully opaque

bdach committed:
Merge branch 'master' into mbd-beatmap-set-cover

bdach commented:
That's what I get for relying on test cases provided... Turns out switching to MBD had introduced a subtle behavioral change in that by default MBD only fades out the old drawable after the new has fully loaded, while previously the cover component did exactly the opposite (old faded out immediately on new load requested). I've fixed this locally for now. Should be better I hope?

frenzibyte commented:
Looks better now, yeah. Sounds like some sort of property that should be added at `ModelBackedDrawable` level for whether to fade in new first then fade out previous, or fade out previous then fade in new, but not worth the time for now until another case pops up or otherwise.

peppy commented on osu.Game/Beatmaps/Drawables/UpdateableBeatmapSetCover.cs:
```diff
@@ -41,6 +41,10 @@ public UpdateableBeatmapSetCover(BeatmapSetCoverType coverType = BeatmapSetCover
 
+        // by default, ModelBackedDrawable hides the old drawable only after the new one has been fully loaded.
+        // this can lead to weird appearance if the cover is not fully opaque, so fade out as soon as a new load is requested in this particular case.
+        protected override void OnLoadStarted() => ApplyHideTransforms(DisplayedDrawable);
```
This change is going to apply everywhere rather than just the place where opacity was an issue. Is that intentional?

bdach replied:
Given that the old code used to work the same way this does (faded out immediately after a `BeatmapSet` set), I think so?

https://github.com/ppy/osu/blob/42c82cea6842b63fb8ce926c081537c4e5214daa/osu.Game/Beatmaps/Drawables/UpdateableBeatmapSetCover.cs#L62-L83

bdach replied:
I mean I *could* disable that behaviour if the cover is opaque, but that seems a bit hacky and weird. Not sure.

frenzibyte replied:
Could make a private subclass of this in that control and let it change the behaviour there instead, since that's where the cover is not opaque if I understand this correctly.

peppy commented:
Breaks beatmap listing overlay:

- Wait for backgrounds to display
- Scroll away and wait a few seconds
- Scroll back

No backgrounds.

peppy reviewed (changes requested):
z

frenzibyte commented:
It looks to have been caused by the transition behaviour override https://github.com/ppy/osu/pull/11582/commits/32b3ea70b966e4087d70d72f1306f55b4fa8b264, it hides the displayed drawable regardless of whether the cause is a model change or just a reload due to masked away (+ should hide the wrapper instead), causing the drawables to not be loaded back at all. (see [here](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Graphics/Containers/ModelBackedDrawable.cs#L139-L164) for how MBD performs the transition)

This can be fixed either, locally, by not applying the hide transform if the model hasn't changed (a bit hacky though, requires a bool that's set when changing the model, then use that in `OnLoadStarted`), or fix at MBD-level, by adding an overridable property that will change the transition to behave the way we want it (hide previous then load new one, rather than load new one then hide previous).

Or a third much nicer way which would be to add the ability to override MBD's behaviour of transitioning wrappers (`ApplyTransition(lastWrapper, currentWrapper)`) and use it on a subclass of `UpdateableBeatmapSetCover` in beatmap listing search control, or with a new `OsuModelBackedDrawable` if the same transition behaviour is reused elsewhere.

@bdach thoughts?

bdach commented:
Actually now that I look again I think the property we want already exists in [`TransformImmediately`](https://github.com/ppy/osu-framework/blob/ab5fa074ac5d5b8d4e496d87a0dc6015d4a321c4/osu.Framework/Graphics/Containers/ModelBackedDrawable.cs#L96-L100)?

frenzibyte commented:
...that really needs a better name don't you think? but yeah it appears to be the one.

bdach commented:
I don't think so, I just didn't read properly.

I tested it, and noticed that in the scenario above (wait for unload and load again) the second load does not have a transition playing. So it's not going to be as simple as just using that I don't think.

frenzibyte commented:
Sounds like https://github.com/ppy/osu-framework/pull/4167?

bdach commented:
jesus christ what a rabbit hole. looks like it, yeah.

frenzibyte committed:
Use `TransformImmediately` instead at beatmap listing search control

Applies two changes:
 - Use `TransformImmediately` which achieves the same wanted transition behaviour without any issues.
 - Move the transition behaviour override into `BeatmapListingSearchControl` in a nested subclass of `UpdateableBeatmapSetCover`.

peppy committed:
Merge branch 'master' into mbd-beatmap-set-cover

peppy commented:
This still looks wrong. The same scenario I mentioned behaves differently on `master` and this PR. With the changes here, scrolling previously loaded panels on screen doesn't show a fade in; skips the preload interval and displays with a one frame flicker.

The point of the logic on master is to avoid rapid texture allocations when scrolling up a list. With this change, it is regressing this behaviour.

frenzibyte commented:
That's a separate issue tracked in https://github.com/ppy/osu-framework/issues/3318 which I've already opened a PR for fixing it here, as it appears to be a simple one: https://github.com/ppy/osu-framework/pull/4167. (that PR really just applies the fade transform regardless of whether the wrapper has changed or not)

peppy commented:
Please re-read my above comment. The issue is not just the missing fade. The timings are changed.

bdach commented:
Yeah this was mentioned in previous comments. The framework PR also linked above is supposed to fix that.

peppy reviewed (changes requested):
As proposed. The timings are changed from `master`.

peppy commented:
In case it isn't obvious:

master:
https://user-images.githubusercontent.com/191335/117398350-22fdff00-af39-11eb-87b5-4283e6c0cec2.mp4


this pr:
https://user-images.githubusercontent.com/191335/117398264-f8ac4180-af38-11eb-9798-8d0ae8f07328.mp4

This is not just the fade timing. I'm saying the *loading* and *web request* timings have changed.



This PR should be blocked or depending on requirements to make this work as expected.

frenzibyte committed:
Add missing load delay

That was a bad one... `ModelBackedDrawable` has a default of 0ms load delay, while previously the wrapper created for beatmap set cover used default 500ms, this change is bringing the load delay back, to avoid firing hundreds of web requests just when doing a quick long scroll on beatmap listing.

frenzibyte commented:
Ah, it was a missing delay before starting the load (was set to `0` as MBD's default), causing thehundreds of web requests fired. Please give it another check, should be fixed now.

peppy commented:
Should this PR not also be marked to depend on https://github.com/ppy/osu-framework/pull/4167?

peppy committed:
Merge branch 'master' into mbd-beatmap-set-cover
